standards/stef_and_luka-standard_proposal_draft.md

```diff
@@ -0,0 +1,274 @@
+    - To solve both, there has to be on-chain code which re-calculates the hash and confirms subscription ID is calculcated with the right parameters and it belongs to the sender. However, it raises a question of why doing it off-chain if on-chain re-calculation has to happen anyway.
+1. Each subscription ID is supplied by provider and standard treats them as black-boxes.
+    - This is the most loose approach and prone to on-chain collisions. It has the same drawbacks as 1.
+1. Each subscription ID is calculated on-chain from a tuple `(providerAddress, externalId)`. External ID can be provider's external subscription ID or hash of subscription parameters or something else entirely.
+    - Moving everything to be done on-chain prevents front-running attacks. We pass subscription parameters to the "create subscription" function and globally unique ID is generated on-chain. Using at least (`providerAddress`, `externalSubscriptionID`) also guarantees there are no ID collisions (NOTE: we rely on the fact that `externalSubscriptionID` supplied by provider is unique).
+
+
+### ERC165 Interface
+
+Standard Interface Detection (ERC165) is used to expose the interfaces that a ERC948 smart contract supports.
+
+### GAS cost and consumption
+
+This proposal supports implementations which can manage arbitrary large number of subscriptions. It avoids defining functions which would require implementations to use for/while loops in the code (e.g. returning all user subscription IDs, etc.). These indicate implementations may not be able to scale and gas costs will rise indefinitely.
+
+
+## ERC948 Interface
+
+```solidity
+  interface ERC948 /* is ERC165 */ {
+
+  /**
+   * @notice User is determined by msg.sender.
+   * @dev Return user subscription providers if there are any.
+   * @param _user User's address.
+   * @return Array of provider addresses.
+   **/
+  function getUserSubscriptionProviders(address _user) external view
+    returns (address[]);
+
+  /**
+   * @notice User is determined by msg.sender.
+   * @dev Return user subscription IDs if there are any.
+   * @param _user User's address.
+   * @param _provider Provider's address.
+   * @return Array of subscription IDs.
+   **/
+  function getUserSubscriptionIds(address _user,
+                                  address _provider) external view
+    returns (uint256[]);
+
+  /**
+   * @dev Return the number of provider subscriptions.
+   * @param _provider Provider's address.
+   * @return Number of subscriptions.
+   **/
+  function getNumberOfProviderSubscriptions(address _provider) external view
```

Comment:
What's the purpose of having this function? Can be easily done using `getUserSubscriptionProviders`.

Reply:
This interface takes into account not only a "contract-per-subscription" model but also a model where one smart contract serves all subscriptions. Additional functions are necessary for such model to be scalable enough to handle millions of subscribers/providers/subscriptions...

Subscription query would then look something like this:
1. Get number of all subscriptions (needed for pagination)
2. Get subscription IDs using pagination
3. Get subscription info

Reply:
As @StefanSimec said - we need to enable arbitrary number of subscriptions per provider. Thus e.g. `getProviderSubscriptionIds` which returns an array of subscription ids is not scalable. Yes, it can handle tens of thousand of elements, maybe even hundred thousand. But at some point the response becomes too large. So the idea here is:

1. Get the number of provider subscriptions.
2. Get a subset of provider subscriptions by passing index and number to the get function. Index is index in the subscription array and number is number of elements starting from the index.

This allows SDKs to easily implement pagination and it solves the scalability problem. And again, we're just building on the shoulders of 721, which has similar requirement.